pkg/kubelet/util/pluginwatcher/plugin_watcher_test.go

```diff
@@ -0,0 +1,170 @@
+	v1beta1 "k8s.io/kubernetes/pkg/kubelet/util/pluginwatcher/example_plugin_apis/v1beta1"
+	v1beta2 "k8s.io/kubernetes/pkg/kubelet/util/pluginwatcher/example_plugin_apis/v1beta2"
+)
+
+func TestExamplePlugin(t *testing.T) {
+	flag.Set("alsologtostderr", fmt.Sprintf("%t", true))
+	var logLevel string
+	flag.StringVar(&logLevel, "logLevel", "4", "test")
+	flag.Lookup("v").Value.Set(logLevel)
+	socketDir, err := ioutil.TempDir("", "plugin_test")
+	require.NoError(t, err)
+	socketPath := socketDir + "/plugin.sock"
+	w := NewWatcher(socketDir)
+
+	testCases := []struct {
+		description string
+		returnErr   error
+	}{
+		{
+			description: "Successfully register plugin through inotify",
+			returnErr:   nil,
+		},
+		{
+			description: "Successfully register plugin through inotify after plugin restarts",
+			returnErr:   nil,
+		},
+		{
+			description: "Fails registration with conflicting plugin name",
+			returnErr:   fmt.Errorf("conflicting plugin name"),
+		},
+		{
+			description: "Successfully register plugin during initial traverse after plugin watcher restarts",
+			returnErr:   nil,
+		},
+		{
+			description: "Fails registration with conflicting plugin name during initial traverse after plugin watcher restarts",
+			returnErr:   fmt.Errorf("conflicting plugin name"),
+		},
+	}
+
+	callbackCount := 0
+	w.AddHandler("example-plugin-type", func(name string, versions []string, sockPath string) error {
+		glog.Infof("receives plugin watcher callback for test: %s\n", testCases[callbackCount].description)
+		require.Equal(t, "example-plugin", name, "Plugin name mismatched!!")
+		require.Equal(t, []string{"v1beta1", "v1beta2"}, versions, "Plugin version mismatched!!")
+		// Verifies the grpcServer is ready to serve services.
+		conn, err := grpc.Dial(socketPath, grpc.WithInsecure(), grpc.WithBlock(),
+			grpc.WithTimeout(10*time.Second),
+			grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+				return net.DialTimeout("unix", addr, timeout)
+			}))
+		require.Nil(t, err)
+		defer conn.Close()
+
+		// The plugin handler should be able to use any listed service API version.
+		v1beta1Client := v1beta1.NewExampleClient(conn)
+		v1beta2Client := v1beta2.NewExampleClient(conn)
+
+		// Tests v1beta1 GetExampleInfo
+		_, err = v1beta1Client.GetExampleInfo(context.Background(), &v1beta1.ExampleRequest{})
+		require.Nil(t, err)
+
+		// Tests v1beta1 GetExampleInfo
+		_, err = v1beta2Client.GetExampleInfo(context.Background(), &v1beta2.ExampleRequest{})
+		require.Nil(t, err)
+		ret := testCases[callbackCount].returnErr
+		callbackCount++
+		return ret
+	})
+	err = w.Start()
+	require.NoError(t, err)
+
+	p := NewTestExamplePlugin() //first cbk
+	go func() {
+		p.Serve(socketPath)
+	}()
+	require.True(t, waitForPluginRegistrationStatus(t, p.registrationStatus))
+
+	// Trying to start a plugin service at the same socket path should fail
+	// with "bind: address already in use"
+	err = p.Serve(socketPath)
```

Comment:
Would be good to have p.Serve() start a go-routine rather than it being blocking.